test/e2e/proxy_test.go

```diff
@@ -1084,3 +1087,171 @@ func testProxyServiceLifeCycle(ipFamily *corev1.IPFamily, ingressIPs []string, d
+// Client IP should always be preserved regardless of whether the traffic is externally or internally originated.
+// Session affinity should take effect when it's enabled.
+func TestProxyLoadBalancerModeDSR(t *testing.T) {
+	skipIfHasWindowsNodes(t)
+	skipIfNumNodesLessThan(t, 3)
+	skipIfAntreaIPAMTest(t)
+	skipIfProxyDisabled(t)
+
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	skipIfProxyAllDisabled(t, data)
+	skipIfLoadBalancerModeIsNot(t, data, config.LoadBalancerModeDSR)
+	skipIfEncapModeIsNot(t, data, config.TrafficEncapModeEncap)
+
+	ingressNode := controlPlaneNodeName()
+	backendNode1 := workerNodeName(1)
+	backendNode2 := workerNodeName(2)
+
+	internalClient := "internal-client"
+	if err := NewPodBuilder(internalClient, data.testNamespace, toolboxImage).OnNode(ingressNode).WithCommand([]string{"sleep", "infinity"}).Create(data); err != nil {
+		t.Fatalf("Failed to create internal client: %v", err)
+	}
+	defer deletePodWrapper(t, data, data.testNamespace, internalClient)
+	var internalClientIPs *PodIPs
+	if internalClientIPs, err = data.podWaitForIPs(defaultTimeout, internalClient, data.testNamespace); err != nil {
+		t.Fatalf("Error when waiting for Pod '%s' to be in the Running state", internalClient)
+	}
+
+	// Create 4 backend Nodes on 2 backend Nodes different from the ingress Node.
+	var wg sync.WaitGroup
+	for i, node := range []string{backendNode1, backendNode1, backendNode2, backendNode2} {
+		wg.Add(1)
+		go func(index int, node string) {
+			defer wg.Done()
+			createAgnhostPod(t, data, fmt.Sprintf("agnhost-%d", index), node, false)
+		}(i, node)
+	}
+	wg.Wait()
+
+	testCases := []struct {
+		name                string
+		isIPv6              bool
+		withSessionAffinity bool
+	}{
+		{
+			name:                "IPv4,withSessionAffinity",
+			isIPv6:              false,
+			withSessionAffinity: true,
+		},
+		{
+			name:                "IPv4,withoutSessionAffinity",
+			isIPv6:              false,
+			withSessionAffinity: false,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.isIPv6 {
+				skipIfNotIPv6Cluster(t)
+			} else {
+				skipIfNotIPv4Cluster(t)
+			}
+			ingressNodeIP := controlPlaneNodeIPv4()
+			ipProtocol := corev1.IPv4Protocol
+			lbIP := "1.1.2.1"
+			externalClientIP := "1.1.1.1"
+			externalClientGateway := "1.1.1.254"
+			externalIPPrefix := 24
+			if tc.isIPv6 {
+				ingressNodeIP = controlPlaneNodeIPv6()
+				ipProtocol = corev1.IPv6Protocol
+				lbIP = "fd75::aabb:ccdd:0201"
+				externalClientIP = "fd75::aabb:ccdd:0101"
+				externalClientGateway = "fd75::aabb:ccdd:01fe"
+				externalIPPrefix = 120
+			}
+
+			// Create another netns to fake an external network on the host network Pod.
+			suffix := randSeq(5)
+			externalClient := fmt.Sprintf("external-client-%s", suffix)
+			externalNetns := fmt.Sprintf("ext-%s", suffix)
+			cmd := fmt.Sprintf(`ip netns add %[1]s && \
```

Comment:
I think we do something similar in several e2e test cases now (e.g., Egress). Any chance to have a shared helper method to generate (and run?) this command?